File: companies/Broadcom/138.copy-list-with-random-pointer.go
package leetcode

// RandomNode138 is the linked-list node with Random pointer for problem 138.
type RandomNode138 struct {
	Val    int
	Next   *RandomNode138
	Random *RandomNode138
}

// CopyRandomList138 interleaves cloned nodes after originals, sets clone random
// pointers via original.random.next, then detaches the clone list. This keeps O(1)
// auxiliary space beyond the output nodes.
//
// Time: O(n). Space: O(1) auxiliary.
func CopyRandomList138(head *RandomNode138) *RandomNode138 {
	if head == nil {
		return nil
	}
	for cur := head; cur != nil; {
		cur.Next = &RandomNode138{Val: cur.Val, Next: cur.Next}
		cur = cur.Next.Next
	}
	for cur := head; cur != nil; cur = cur.Next.Next {
		if cur.Random != nil {
			cur.Next.Random = cur.Random.Next
		}
	}
	copiedHead := head.Next
	for cur := head; cur != nil; {
		copied := cur.Next
		cur.Next = copied.Next
		cur = cur.Next
		if cur != nil {
			copied.Next = cur.Next
		} else {
			copied.Next = nil
		}
	}
	return copiedHead
}

File: VMware/138.copy-list-with-random-pointer.py
"""
Approach: Interleave copied nodes with originals, assign random pointers, then detach.
Data structure: the list itself temporarily acts as the original-to-copy map because each copy is original.next.
Interview logic: after weaving A -> A' -> B -> B', the clone of cur.random is cur.random.next. Once random pointers are assigned, restore original next links and extract the copied list.
Complexity: O(n) time, O(1) extra space excluding output nodes.
Tests and edge cases: empty list returns None; random may be None; self-random and random cycles copy correctly.
"""
from __future__ import annotations
from typing import Optional

# @lc code=start
"""
# Definition for a Node.
class Node:
    def __init__(self, x: int, next: 'Node' = None, random: 'Node' = None):
        self.val = int(x)
        self.next = next
        self.random = random
"""
class Solution:
    def copyRandomList(self, head: 'Optional[Node]') -> 'Optional[Node]':
        if not head:
            return None
        cur = head
        while cur:
            cur.next = Node(cur.val, cur.next)
            cur = cur.next.next
        cur = head
        while cur:
            if cur.random:
                cur.next.random = cur.random.next
            cur = cur.next.next
        cur = head
        copied_head = head.next
        while cur:
            copied = cur.next
            cur.next = copied.next
            cur = cur.next
            copied.next = cur.next if cur else None
        return copied_head
# @lc code=end
